app/javascript/packages/manageable-authenticator/manageable-authenticator-element.ts

```diff
@@ -0,0 +1,242 @@
+  }
+
+  get doneButton(): HTMLButtonElement {
+    return this.querySelector<HTMLButtonElement>('.manageable-authenticator__done-button')!;
+  }
+
+  get editContainer(): HTMLElement {
+    return this.querySelector<HTMLElement>('.manageable-authenticator__edit')!;
+  }
+
+  get renameButton(): HTMLButtonElement {
+    return this.querySelector<HTMLButtonElement>('.manageable-authenticator__rename-button')!;
+  }
+
+  get deleteButton(): SpinnerButtonElement {
+    return this.querySelector<SpinnerButtonElement>('.manageable-authenticator__delete-button')!;
+  }
+
+  get renameInput(): HTMLInputElement {
+    return this.querySelector<HTMLInputElement>('.manageable-authenticator__rename-input')!;
+  }
+
+  get renameForm(): HTMLFormElement {
+    return this.querySelector<HTMLFormElement>('.manageable-authenticator__rename')!;
+  }
+
+  get saveRenameButton(): SpinnerButtonElement {
+    return this.querySelector<SpinnerButtonElement>(
+      '.manageable-authenticator__save-rename-button',
+    )!;
+  }
+
+  get cancelRenameButton(): HTMLButtonElement {
+    return this.querySelector<HTMLButtonElement>(
+      '.manageable-authenticator__cancel-rename-button',
+    )!;
+  }
+
+  get alert(): HTMLElement {
+    return this.querySelector<HTMLElement>('.manageable-authenticator__alert')!;
+  }
+
+  get uniqueId(): string {
+    return this.getAttribute('unique-id')!;
+  }
+
+  get apiURL(): string {
+    return this.getAttribute('api-url')!;
+  }
+
+  get reauthenticationURL(): string {
+    return this.getAttribute('reauthentication-url')!;
+  }
+
+  get reauthenticateAt(): Date {
+    return new Date(this.getAttribute('reauthenticate-at')!);
+  }
+
+  get name(): string {
+    return this.getAttribute('configuration-name')!;
+  }
+
+  set name(name: string) {
+    this.setAttribute('configuration-name', name);
+
+    this.querySelectorAll('.manageable-authenticator__name').forEach((nameElement) => {
+      nameElement.textContent = name;
+    });
+  }
+
+  toggleEditing(isEditing: boolean) {
+    this.toggleRenaming(false);
+
+    this.classList.toggle('manageable-authenticator--editing', isEditing);
+
+    const focusTarget = isEditing ? this.editContainer : this.manageButton;
+    focusTarget.focus();
+
+    this.#setAlertMessage(null);
+  }
+
+  toggleRenaming(isRenaming: boolean) {
+    this.classList.toggle('manageable-authenticator--renaming', isRenaming);
+
+    this.renameInput.value = this.name;
+
+    const focusTarget = isRenaming ? this.renameInput : this.editContainer;
+    focusTarget.focus();
+
+    if (isRenaming) {
+      this.renameInput.setSelectionRange(
+        this.renameInput.value.length,
+        this.renameInput.value.length,
+      );
+    }
+
+    this.#setAlertMessage(null);
+  }
+
+  async delete() {
+    // Disable reason: This is an intentional user-facing confirmation prompt.
+    /* eslint-disable-next-line no-alert */
+    if (!window.confirm(this.strings.deleteConfirm)) {
```

Comment:
I see we have a few specs that do Capybara's `accept_confirm`, do we need ones that do `dismiss_confirm`?

And it looks like we're not testing/stubbing `window.confirm` in JS tests? Is the idea that the acceptance specs cover the full flow, and that the work to maintain the DOM/HTML markup in a JS-only test would be not worth it?

Reply:
> I see we have a few specs that do Capybara's `accept_confirm`, do we need ones that do `dismiss_confirm`?

Yeah, it'd probably be good to include coverage for that somewhere, maybe in an existing spec for performance/simplicity's sake, since we could do something like...

1. Click delete
2. Dismiss confirmation
3. Assert cancellation expectations
4. Click delete
5. Accept confirmation
6. Assert confirmation expectations

>And it looks like we're not testing/stubbing window.confirm in JS tests? Is the idea that the acceptance specs cover the full flow, and that the work to maintain the DOM/HTML markup in a JS-only test would be not worth it?

It is being stubbed:

https://github.com/18F/identity-idp/blob/6ee9839d8306b3577cd05bdefd6db329043bc93b/app/javascript/packages/manageable-authenticator/manageable-authenticator-element.spec.ts#L353

https://github.com/18F/identity-idp/blob/6ee9839d8306b3577cd05bdefd6db329043bc93b/app/javascript/packages/manageable-authenticator/manageable-authenticator-element.spec.ts#L378

https://github.com/18F/identity-idp/blob/6ee9839d8306b3577cd05bdefd6db329043bc93b/app/javascript/packages/manageable-authenticator/manageable-authenticator-element.spec.ts#L412

https://github.com/18F/identity-idp/blob/6ee9839d8306b3577cd05bdefd6db329043bc93b/app/javascript/packages/manageable-authenticator/manageable-authenticator-element.spec.ts#L439

(If anything, it was the other way around, I was lax on complete feature spec coverage re: the dismiss case, since I knew the JavaScript tests would have more thorough coverage)

Reply:
ah I missed the `sandbox.stub` calls in my search, that's probably fine/close enough

Reply:
I did add a bonus assertion in 0c100fd to cover the dismiss flow in feature specs.